clippy_lints/src/strings.rs

```diff
@@ -173,16 +176,63 @@ fn is_add(cx: &LateContext<'_>, src: &Expr<'_>, target: &Expr<'_>) -> bool {
     }
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Check if the string is transformed to byte array and casted back to string.
+    ///
+    /// **Why is this bad?** It's unnecessary, the string can be used directly.
+    ///
+    /// **Known problems:** None
+    ///
+    /// **Example:**
+    /// ```rust
+    /// let string = "Hello World!";
+    /// let string2 = std::str::from_utf8(&string.as_bytes()[6..11]).unwrap();
+    /// ```
+    /// could be written as
+    /// ```rust
+    /// let string = "Hello World!";
+    /// let string2 = &string[6..11];
+    /// ```
+    pub STRING_FROM_UTF8_AS_BYTES,
+    correctness,
+    "make sure that access from slice should be done from bytes and not from string directly"
+}
+
 // Max length a b"foo" string can take
 const MAX_LENGTH_BYTE_STRING_LIT: usize = 32;
 
-declare_lint_pass!(StringLitAsBytes => [STRING_LIT_AS_BYTES]);
+declare_lint_pass!(StringLitAsBytes => [STRING_LIT_AS_BYTES, STRING_FROM_UTF8_AS_BYTES]);
 
 impl<'tcx> LateLintPass<'tcx> for StringLitAsBytes {
     fn check_expr(&mut self, cx: &LateContext<'tcx>, e: &'tcx Expr<'_>) {
         use crate::utils::{snippet, snippet_with_applicability};
         use rustc_ast::LitKind;
 
+        if_chain! {
+            // Find std::str::converts::from_utf8
+            if let Some(args) = match_function_call(cx, e, &paths::STR_FROM_UTF8);
+
+            // Find string::as_bytes
+            if let ExprKind::AddrOf(BorrowKind::Ref, _, ref args) = args[0].kind;
+            if let ExprKind::Index(ref left, ref right) = args.kind;
+            let (method_names, _, _) = method_calls(left, 1);
+            if method_names.len() == 1;
+            if method_names[0] == sym!(as_bytes);
```

Comment:
I think this might be `if &[sym!(as_bytes)] == method_names;`.

Reply:
This is not possible since `method_names` is a vector and the other side is a fixed size array `&[something, 1]`, I may be wrong, but I was unable to find a way to accomplish such thing.

Reply:
Hmmmm...perhaps `&method_names` or even `&method_names[..]`? Sorry, I don't have the time to check.

Reply:
If the intention is to use a slice pattern, it should be an `if let`